hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/health/MismatchedReplicasHandler.java

```diff
@@ -57,35 +58,28 @@ public MismatchedReplicasHandler(
    */
   @Override
   public boolean handle(ContainerCheckRequest request) {
-    ContainerInfo containerInfo = request.getContainerInfo();
-    Set<ContainerReplica> replicas = request.getContainerReplicas();
+    if (request.isReadOnly()) {
+      return false;
+    }
+
+    final ContainerInfo containerInfo = request.getContainerInfo();
+    final Set<ContainerReplica> replicas = request.getContainerReplicas();
+
     if (containerInfo.getState() != HddsProtos.LifeCycleState.CLOSED &&
         containerInfo.getState() != HddsProtos.LifeCycleState.QUASI_CLOSED) {
       // Handler is only relevant for CLOSED or QUASI-CLOSED containers.
       return false;
     }
-    LOG.debug("Checking container {} in MismatchedReplicasHandler",
-        containerInfo);
 
-    if (request.isReadOnly()) {
-      return false;
-    }
-    // close replica if needed
-    for (ContainerReplica replica : replicas) {
-      if (shouldBeClosed(containerInfo, replica)) {
-        LOG.debug("Sending close command for mismatched replica {} of " +
-            "container {}.", replica, containerInfo);
+    LOG.debug("Checking container {} in MismatchedReplicasHandler", containerInfo);
 
-        if (containerInfo.getState() == HddsProtos.LifeCycleState.CLOSED) {
-          replicationManager.sendCloseContainerReplicaCommand(
-              containerInfo, replica.getDatanodeDetails(), true);
-        } else if (containerInfo.getState() ==
-            HddsProtos.LifeCycleState.QUASI_CLOSED) {
-          replicationManager.sendCloseContainerReplicaCommand(
-              containerInfo, replica.getDatanodeDetails(), false);
-        }
-      }
-    }
+    final Predicate<ContainerReplica> shouldSendClose = (r) -> shouldSendClose(containerInfo, r);
+
+    replicas.stream().filter(shouldSendClose).forEach(r -> {
+      LOG.debug("Sending close command for mismatched replica {} of container {}.", r, containerInfo);
+      replicationManager.sendCloseContainerReplicaCommand(
+          containerInfo, r.getDatanodeDetails(), shouldForceClose(containerInfo, r));
```

Comment:
So just to be clear, we're also sending a force close if the replica is open/closing and the bcsid matches. I think this is different from the previous commit's behaviour.

Reply:
Yes, if the BCSID matches we are sending force close command to that Container replica.

This is not different from the previous behaviour, we were sending force close for open/closing replica even without checking the sequence Id. With this change we will only send force close if the sequence Id matches.

We believe  that the BCSID of a CLOSED container is the highest (correct).

Reply:
No, I meant it's different from the behaviour that @swamirishi had pushed in his commit. Your latest one looks fine to me, but @swamirishi should take a look.

Reply:
@nandakumar131 Lets not send a force close directly from OPEN/CLOSING. Let us move it to QUASI_CLOSED and let the QUASI_CLOSED container handler do it. Let us not have duplicate coding logic floating around.

Reply:
yeah i agree with you @errose28

Reply:
The same handler is used for EC Containers as well and for EC Containers the Datanode [expects force flag to be set](https://github.com/apache/ozone/blob/39437eae0cbc4de6078ca16f80d5c579d5a93fbf/hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/commandhandler/CloseContainerCommandHandler.java#L128) for Replicas in OPEN/CLOSING state.

We should add logic based on Container type.

If everyone feels strongly about this, I can change the behaviour for Ratis Containers.